sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -158,49 +151,67 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
-            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
-            requests.add(request);
+        if (strategy instanceof ThresholdBasedAvailabilityStrategy) {
+            List<URI> effectiveEndpoints = getApplicableEndPoints(request);
+            if (effectiveEndpoints != null) {
+                effectiveEndpoints
+                    .forEach(locationURI -> {
+                        if (locationURI != null) {
+                            RxDocumentServiceRequest newRequest = request.clone();
+                            newRequest.requestContext.routeToLocation(locationURI);
+                            if (monoList.isEmpty()) {
+                                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout));
+                            } else {
+                                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout)
+                                    .delaySubscription(((ThresholdBasedAvailabilityStrategy) strategy).getThreshold()
+                                        .plus(((ThresholdBasedAvailabilityStrategy) strategy)
+                                            .getThresholdStep().multipliedBy(monoList.size() - 1))));
+                            }
+                        }
+                    });
+            }
         }
 
-        for (URI endpoint : speculativeProcessor.getRegionsToSpeculate(config, this.transportClient
-            .getGlobalEndpointManager().getReadEndpoints())) {
-            if (request.requestContext.locationEndpointToRoute != endpoint) {
-                RxDocumentServiceRequest newRequest = request.clone();
-                newRequest.requestContext.routeToLocation(endpoint);
-                requests.add(newRequest);
-                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout)
-                    .delaySubscription(speculativeProcessor.getThreshold(config).plus(speculativeProcessor.getThresholdStepDuration(config, monoList.size() - 1))));
-            }
+        // If the above conditions are not met, then we will just return the original request
+        if (monoList.isEmpty()) {
+            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
         }
 
-        return Mono.firstWithValue(monoList).map(storeResponse ->
-            {
-                for (RxDocumentServiceRequest r : requests) {
-                    CosmosDiagnostics diagnostics = r.requestContext.cosmosDiagnostics;
-                    // for the successful request region, we will use the actual latency
-                    if (r.getActivityId().toString().equals(storeResponse.getActivityId())) {
-                        speculativeProcessor.onResponseReceived(r.requestContext.locationEndpointToRoute,
-                            diagnostics.getDuration());
-                    }
-                }
-                return storeResponse;
-            });
+        return Mono.firstWithValue(monoList);
```

Comment:
Do we need to merge diagnostics here?

Reply:
they share the diagnostics context so should be good